client/src/controllers/RulesController.ts

```diff
@@ -0,0 +1,156 @@
+ * along with a set of rules to match against that data.
+ *
+ * @example - Enable a button if a value is chosen
+ * <form data-controller="w-rules" data-action="change->w-rules#resolve">
+ *   <select name="fav-drink" required>
+ *     <option value="">Select a drink</option>
+ *     <option value="coffee">Coffee</option>
+ *     <option value="other">Other</option>
+ *   </select>
+ *   <button type="button" data-w-rules-target="enable" data-rule='{"fav-drink": ["coffee"]}'>
+ *     Continue
+ *   </button>
+ * </form>
+ *
+ */
+export class RulesController extends Controller<HTMLFormElement> {
+  static targets = ['enable'];
+
+  /** Targets will be enabled if the `data-rule` matches the scoped form data, otherwise will be disabled. */
+  declare readonly enableTargets: HTMLElement[];
+  declare readonly hasEnableTarget: boolean;
+
+  declare active: boolean;
+  declare ruleCache: Record<string, Record<string, string[]>>;
+
+  initialize() {
+    this.ruleCache = {};
+    this.resolve = debounce(this.resolve.bind(this), 50);
+  }
+
+  connect() {
+    this.checkTargets();
+  }
+
+  /**
+   * Checks for any targets that will mean that the controller needs to be active.
+   */
+  checkTargets() {
+    this.active = this.hasEnableTarget;
+  }
+
+  /**
+   * Resolve the conditional targets based on the form data and the target(s)
+   * `data-rule` attributes.
+   */
+  resolve() {
+    if (!this.active) return;
+
+    const form = this.element;
+    const formData = Object.fromEntries(new FormData(form).entries());
+
+    [
+      ...this.enableTargets.map((target) => ({ shouldDisable: false, target })),
+    ].forEach(({ shouldDisable, target }) => {
+      this.toggleAttribute(
+        target,
+        this.getIsMatch(formData, this.getRuleData(target))
+          ? !shouldDisable
+          : shouldDisable,
+        'disabled',
+      );
+    });
+
+    this.dispatch('resolved', { bubbles: true, cancelable: false });
+  }
+
+  getIsMatch(
+    formData: Record<string, FormDataEntryValue>,
+    ruleData: Record<string, string[]>,
+  ): boolean {
+    return (
+      ruleData &&
+      Object.entries(ruleData).every(([key, value]) =>
+        value.includes(String(formData[key] || '')),
+      )
+    );
+  }
+
+  getRuleData(target: Element): Record<string, string[]> {
+    if (!target) return {};
+    const ruleStr = target.getAttribute('data-rule');
+    if (!ruleStr) return {};
+
+    // check cache
+    const cachedRule = this.ruleCache[ruleStr];
+    if (cachedRule) return cachedRule;
+
+    // prepare parsed rule data
+    let rule = {};
+
+    if (ruleStr) {
+      try {
+        rule = JSON.parse(ruleStr);
+        if (Array.isArray(rule)) rule = Object.fromEntries(rule);
+      } catch (e) {
+        // Safely ignore JSON parsing errors
+      }
+    }
+
+    // Map through values and convert to array of strings
+    // Allowing falsey values to be treated as an empty string
```

Comment:
My goal here is to allow for `data-rule='{"fav-drink": false}'` to trigger when the field with name `fav-drink` is has a blank (empty string) value. This was me thinking about how it would be intuitive to write these rules.

However, I see your point. We could assume that anyone writing these rules would know that form values can only be strings, but then using `false`, `null`, or `undefined` would basically never work at all.

Unless we do coerce matching, but I feel this makes the complexity in the matcher too much.

For now, I will just remove this smartness, we will trust devs to know that an empty string is the only way to check for 'falsey' values.

Reply:
This has been updated.